src/core/fs.rs

```diff
@@ -117,3 +121,246 @@ pub fn set_group(path: &Path, group: &str) -> Result<(), io::Error> {
+    file: PathBuf,
+    read_fd: RawFd,
+    write_fd: RawFd,
+    state: Mutex<Option<FileWatcherState>>,
+}
+
+pub struct FileWatcher {
+    data: Arc<FileWatcherData>,
+}
+
+#[derive(Debug)]
+pub struct FileWatcherError;
+
+impl FileWatcher {
+    pub fn new<P: AsRef<Path>>(file_path: P) -> Result<Self, FileWatcherError> {
+        let file = file_path.as_ref();
+
+        let dir = match file.parent() {
+            Some(p) => p,
+            None => return Err(FileWatcherError),
+        };
+
+        let mut fds = [0; 2];
+
+        // SAFETY: fds pointer is valid
+        let ret = unsafe { libc::pipe(fds.as_mut_ptr()) };
+        assert_eq!(ret, 0);
+
+        for fd in &fds {
+            set_fd_nonblocking(*fd).unwrap();
+        }
+
+        let data = Arc::new(FileWatcherData {
+            file: file.to_owned(),
+            read_fd: fds[0],
+            write_fd: fds[1],
+            state: Mutex::new(None),
+        });
+
+        let watcher = {
+            let data = Arc::clone(&data);
+
+            notify::recommended_watcher(move |event: Result<notify::Event, notify::Error>| {
+                let event = match event {
+                    Ok(event) => event,
+                    Err(e) => {
+                        warn!("file watcher error: {:?}", e);
+                        return;
+                    }
+                };
+
+                if !event.paths.into_iter().any(|p| p == data.file) {
+                    // skip unrelated events
+                    return;
+                }
+
+                if let Some(state) = data.state.lock().unwrap().as_mut() {
+                    if !state.changed {
+                        state.changed = true;
+
+                        // non-blocking write to wake up the other side
+
+                        let buf: [u8; 1] = [0; 1];
+
+                        // SAFETY: buf pointer and size are valid
+                        let ret = unsafe {
+                            libc::write(data.write_fd, buf.as_ptr() as *const libc::c_void, 1)
+                        };
+                        assert!(ret == 1 || get_errno() == libc::EAGAIN);
+                    }
+                }
+            })
+            .expect("failed to create file watcher")
+        };
+
+        {
+            let mut state = data.state.lock().unwrap();
+
+            let state = state.insert(FileWatcherState {
+                watcher,
+                changed: false,
+            });
+
+            // watch the dir instead of the file, so we can detect file creates
+            if let Err(e) = state
+                .watcher
+                .watch(dir, notify::RecursiveMode::NonRecursive)
+            {
+                warn!("failed to watch {}: {:?}", dir.display(), e);
+            }
+        }
+
+        Ok(Self { data })
+    }
+
+    pub fn file_changed(&self) -> bool {
+        let mut changed = false;
+
+        if let Some(state) = self.data.state.lock().unwrap().as_mut() {
```

Comment:
Same Nitpick: state lock unwrap that should at least be an .expect() to help if this ever goes wrong.